x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindFieldStructureAction.java

```diff
@@ -66,11 +80,16 @@ protected void doExecute(Task task, FindFieldStructureAction.Request request, Ac
                     );
                     return;
                 }
-                List<String> messages = getMessages(searchResponse, request.getField());
+
+                // As matching a regular expression might take a while, we run in a different thread to avoid blocking the network thread.
+                var runnable = ActionRunnable.supply(listener, () -> {
+                    var messages = getMessages(searchResponse, request.getField());
+                    return buildTextStructureResponse(messages, request);
+                });
                 try {
-                    listener.onResponse(buildTextStructureResponse(messages, request));
-                } catch (Exception e) {
-                    listener.onFailure(e);
+                    threadPool.generic().execute(runnable);
+                } catch (RejectedExecutionException e) {
+                    runnable.onRejection(e);
                 }
             }, listener::onFailure));
```

Comment:
nit: use `ActionListener.delegateFailureAndWrap` rather than a bare `ActionListener.wrap` that manually delegates failures like this.